x/leverage/keeper/keeper.go

```diff
@@ -274,33 +274,43 @@ func (k Keeper) Repay(ctx sdk.Context, borrowerAddr sdk.AccAddress, payment sdk.
 }
 
 // Collateralize enables selected uTokens for use as collateral by a single borrower.
-func (k Keeper) Collateralize(ctx sdk.Context, borrowerAddr sdk.AccAddress, coin sdk.Coin) error {
-	if err := k.validateCollateralize(ctx, coin); err != nil {
+func (k Keeper) Collateralize(ctx sdk.Context, borrowerAddr sdk.AccAddress, uToken sdk.Coin) error {
+	if err := k.validateCollateralize(ctx, uToken); err != nil {
 		return err
 	}
 
-	currentCollateral := k.GetCollateralAmount(ctx, borrowerAddr, coin.Denom)
-	if err := k.setCollateralAmount(ctx, borrowerAddr, currentCollateral.Add(coin)); err != nil {
+	max, err := k.maxCollateral(ctx, uToken.Denom)
+	if err != nil {
 		return err
 	}
 
-	return k.bankKeeper.SendCoinsFromAccountToModule(ctx, borrowerAddr, types.ModuleName, sdk.NewCoins(coin))
+	total := k.GetTotalCollateral(ctx, uToken.Denom)
```

Comment:
`k.maxCollateral` already computes that, would be good to do it once - maybe we can return it from `k.maxCollateral`?

Reply:
There are a few efficiency upgrades we can do here (and also in the market summary query) but I'm not sure how far we should optimize in this PR.

Another option would be to make a function `k.CollateralShare(denom)` and use it in an error check at the _end_ of `k.Collateralize`, while continuing to use function `maxCollateral` in `grpc_query.go`.

I ended up avoiding that optimization to keep the code smaller, but would prefer it if we have to optimize now.

Reply:
Refactored to removed duplicate